### PR TITLE
STT Support IOS [TDH-255]

### PR DESCRIPTION
**Pull request by Sanjeevsinghnegi**
### What do you want to achieve?
- Fix Speech to Text not working on IOS devices and Mac Safari
- This code changes only for Apple devices (Safari browser)
- User can add ` voiceInputTimeout: <IN SEC>` for checking the timeout before closing STT (Safari only)

### PR Checklist

- [x] I have tested it locally and all functionalities are working fine.
- [ ] I have compared it with mocks and all design elements are the same.
- [ ] I have tested it in IE Browser.

### How was the code tested?

- Running ngrock and deploying this branch on widget-test
- Checked on Mac > Safari, chrome, edge
- Checked on Windows > Chrome, edge
- Checked on IOS > Safari, chrome
- Checked on Android > Chrome

### What new thing you came across while writing this code? 
- So, STT is not working properly on Safari, we have added some conditions to get it working for Safari

### Screen Recording

https://user-images.githubusercontent.com/66241555/233826860-d73fbc8e-f659-4adf-9a82-eed57fc5c432.mov

NOTE: Make sure you're comparing your branch with the correct base branch

### CODE REVIEW

**Review by prochnost04 — APPROVED** [review]

**Review by panwaranuj01 — APPROVED** [review]

**Review by panwaranuj01 — APPROVED** [review]

File: webplugin/js/app/media/media-service.js [hunk 1]
```diff
@@ -1,6 +1,28 @@
...
-       window.navigator.language || window.navigator.userLanguage || 'en-us',
+        window.navigator.language || window.navigator.userLanguage || 'en-US',
+    isAppleDevice: function () {
+        var isIOSDevice = /iPhone|iPad|iPod/i.test(navigator.userAgent);
+        var isMacSafari = /^((?!chrome|android).)*safari/i.test(navigator.userAgent);
+
+        return isIOSDevice || isMacSafari
+    },
+    endSttExplicitly: function (lastListeningEventTime, recognizingDone, recognition) {
+        const checkTimeSTT = 1000;
```
Comment (devashishmamgain) [inline webplugin/js/app/media/media-service.js:11]: what is checkTimeSTT?

File: webplugin/js/app/media/media-service.js [hunk 2]
```diff
@@ -10,23 +32,31 @@ Kommunicate.mediaService = {
     processVoiceInputClickedEvent: function () {
         kmWidgetEvents.eventTracking(eventMapping.onVoiceIconClick);
         if (!('webkitSpeechRecognition' in window)) {
-            alert('browser do not support speech recogization');
+            alert('browser do not support speech recognition');
         } else {
-            var recognition = new webkitSpeechRecognition();
+            //As of April 2023, works only in chrome, edge and safari(limited support)
+            var recognizingDone = false;
+            var lastListeningEventTime = null;
```
Comment (devashishmamgain) [inline webplugin/js/app/media/media-service.js:39]: this is declared locally here but used outside too

File: webplugin/js/app/media/media-service.js (hunk 1 above)
Comment (Sanjeevsinghnegi) [reply webplugin/js/app/media/media-service.js:11]: need to change the variable name, earlier it was used as a interval time as well as for millisecond to sec conversion
updated it to
const MILLISECOND_TO_SEC = 1000;

File: webplugin/js/app/media/media-service.js [hunk 3]
```diff
@@ -1,6 +1,28 @@
...
+        window.navigator.language || window.navigator.userLanguage || 'en-US',
+    isAppleDevice: function () {
+        var isIOSDevice = /iPhone|iPad|iPod/i.test(navigator.userAgent);
+        var isMacSafari = /^((?!chrome|android).)*safari/i.test(navigator.userAgent);
+
+        return isIOSDevice || isMacSafari
+    },
+    endSttExplicitly: function (lastListeningEventTime, recognizingDone, recognition) {
+        const checkTimeSTT = 1000;
+        const customTimeSet = 5; //To do setting this via widget script
```
[the remark this answers is not in the document]
Comment (Sanjeevsinghnegi) [reply webplugin/js/app/media/media-service.js:12]: this is the time for which the microphone listens to any input if not receive any input then after 5 sec it stops

in chrome, as soon as the user stops speaking within 1-sec end event fires, and the microphone off
you can check this on over prod website

File: webplugin/js/app/media/media-service.js (hunk 2 above)
Comment (Sanjeevsinghnegi) [reply webplugin/js/app/media/media-service.js:39]: yes this is explicitly used in the set interval function to get the last time for which user speak

File: webplugin/js/app/media/media-service.js [hunk 4]
```diff
@@ -1,6 +1,37 @@
...
+        return isIOSDevice || isMacSafari;
+    },
+    endSttExplicitly: function (
+        lastListeningEventTime,
+        recognizingDone,
+        recognition,
+        checkLastSpeech
+    ) {
+        const MILLISECOND_TO_SEC = 1000;
+        const CUSTOM_TIMEOUT_STT = 5; //To do set this via widget script
```
Comment (devashishmamgain) [inline webplugin/js/app/media/media-service.js:19]: move it to app option settings

File: webplugin/js/app/media/media-service.js [hunk 5]
```diff
@@ -41,39 +81,54 @@ Kommunicate.mediaService = {
...
                 window.$applozic.fn.applozic('toggleMediaOptions');
             };
+            recognition.onerror = function (event) {
+                console.log('error while speech recognition', event.error);
+                recognition.abort();
+            };
+
+            //explicitly Stop the Mic recording only for Safari
+            if (isAppleProduct) {
+                var checkLastSpeech = setInterval(function () {
```
Comment (devashishmamgain) [inline webplugin/js/app/media/media-service.js:97]: will this keep running even if user is not using voice?

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/media/media-service.js (hunk 5 above)
Comment (Sanjeevsinghnegi) [reply webplugin/js/app/media/media-service.js:97]: no, 
1> It only runs whenever the user has trying to access the widget speech-to-text function from an Apple device
2>  once it enables, it continues to run every 1 sec until the condition matches then we cleared this interval hence set interval stops

Comment (Sanjeevsinghnegi) [reply webplugin/js/app/media/media-service.js:97]: Uploading Screen Recording 2023-04-23 at 1.08.17 PM.mov…

File: webplugin/js/app/media/media-service.js (hunk 4 above)
Comment (Sanjeevsinghnegi) [reply webplugin/js/app/media/media-service.js]: Done, getting this from widget script also by default time is 10sec

**Review by prochnost04 — APPROVED** [review]

**Review by devashishmamgain — APPROVED** [review]

**Pull request merged**